### ISSUE TITLE
Uncaught TypeError: Cannot read property '$wire' of undefined

### ISSUE THREAD

**Issue by elfeffe** [livewire/livewire#1790]
I have installed Livewire using TALL preset in Laravel 7.
I have upgraded it to Livewire 2 and Laravel 8.

I'm having some problems like getting Uncaught TypeError: Cannot read property '$wire' of undefined when I try to use

`@this.model = 'test;
`
(model is the name of my property)

I have tried using Alpine in the HEAD from CDN, and using Alpine from my bundle (the TALL preset installs it there).
Did I forget something when upgrading?

I have seen this too:
https://forum.laravel-livewire.com/t/cannot-read-property-wire-of-undefined/1678

Is similar, and he says how to resolve the issue, but I can't

**Comment by elfeffe** [comment]
I think I have found the problem.
I use             

`$livewireBlock = Livewire::mount($block['name'], $properties);
`

To load my components (because I'm working on a system that loads the components dynamically).

Is there any way to pass the KEY int hat case? Hoe to set the KEY or get the ID if I don't use Blade to load my components?

**Comment by austenc** [comment]
@elfeffe 

Are you loading Alpine _after_ Livewire? Assuming you've tried this since you referenced the forum thread.

It looks like the `Livewire::mount` method is the one responsible for generating the component ID, see here:
https://github.com/livewire/livewire/blob/master/src/LivewireManager.php#L76

Could you provide more information about why you're trying to load components dynamically or what you're trying to accomplish? Perhaps with some more information, it might be easier to help.

**Comment by elfeffe** [comment]
I'm making a builder, and I load some components dynamically based on the block that the user has selected.
It works great, but it gives some small problems that I'm handling with AlpineJS.
It doesn't work because @this doesn't return an ID if I mount it in my way. But the truth is that I didn't try LW2 in my project using Blade yet, so maybe I'm having the same problem there.
What I exactly need is to set a value in my model using Javascript, because I'm using https://github.com/UniSharp/laravel-filemanage for image uploads, and it returns the URL that then I have to pass to my model using plain JS.
I'm also trying to do it with alpine (first I dispatch an event that I try to catch using Alpine to then set the model using $wire).
Is there any way to set a property using JS that is not the one below?

`        @this.count = 5
`

**Comment by elfeffe** [comment]
@austenc 

I'm doing it in a different way now.

I adding my dynamic  components using this:

 ```
@foreach($form_fields as $key => $field)
        @livewire($field['component'], $field['properties'], key($key)) 
@endforeach
```

Where $field['component'] is the name of the component and the properties are located in $field['properties']
That is included in a component named "Edit". And it is working.

Now inside one of those components (child of Edit), I have this:
 
```
@image-url.window="$wire.setOriginalUrl($event.detail.url)"`
 ```

So I catch the event **image-url** and try to call to a function inside my child component.
But I get an error because Edit (which is the parent), doesn't have that function.

Is it the normal behavior or it's a bug?

**Comment by elfeffe** [comment]
@image-url.window="$wire.setOriginalUrl($event.detail.url)"`+
That code is inside the child component, not the parent.
That's why I think it should fire the correct component, but it's firing the parent.
I will continue the investigation

**Comment by jonneroelofs** [comment]
I had this earlier today because my child component had multiple root elements. When I wrapped all the html inside the child component inside a <div> it started calling my the function on the child livewire component instead of the parent.

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by hazem-taha** [comment]
m still having this issue, any chance ?

**Comment by elfeffe** [comment]
> m still having this issue, any chance ?

I think it's related to a component inside another component. I'm not sure about how to resolve this, but I think the problem is with recursive components.
Check above:

> @elfeffe is the @image-url stuff on the child component itself, or within its template? The $wire variable will only reference the current context, so you'll need that within the child component's template for it to work afaik.

This can help

**Comment by hazem-taha** [comment]
> > m still having this issue, any chance ?
> 
> I think it's related to a component inside another component. I'm not sure about how to resolve this, but I think the problem is with recursive components.
> Check above:
> 
> > @elfeffe is the @image-url stuff on the child component itself, or within its template? The $wire variable will only reference the current context, so you'll need that within the child component's template for it to work afaik.
> 
> This can help

I tried calling `$wire` from an alpinejs component in a wire component without nesting but still getting the error, although if i call `@this.SOME_ACTION` , it works.

**Comment by joshhanley** [comment]
Yeah there are a few issues with `$wire`, which I have submitted a fix for, see this PR alpinejs/alpine#1160

So hopefully will be fixed soon!

Hope this helps!
